packages/shared/src/components/squads/SquadActionButton.tsx

```diff
@@ -113,12 +113,61 @@ export const SquadActionButton = ({
     squad.currentMember?.role === SourceMemberRole.Blocked;
   const isCurrentMember = !!squad.currentMember && !isMemberBlocked;
 
+  const fuzzyQueryKey = generateQueryKey(
+    RequestKey.Sources,
+    null,
+    undefined,
+    true,
+    squad.category?.id,
+  );
+  const fuzzyQueryMatch = queryClient.getQueriesData({
+    queryKey: fuzzyQueryKey,
+  });
+  const fuzzyFeaturedQueryMatch = queryClient.getQueriesData({
+    queryKey: generateQueryKey(RequestKey.Sources, null, true, true),
+  });
+  const categoryQueryKey = fuzzyQueryMatch?.[0]?.[0];
+  const featuredQueryKey = fuzzyFeaturedQueryMatch?.[0]?.[0];
+
+  const joinSquadMutation = (data) => {
+    return {
+      ...data,
+      pages: data?.pages?.map((edge) => {
+        return {
+          ...edge,
+          sources: {
+            ...edge.sources,
+            edges: edge.sources.edges.map((subEdge) => {
+              const { node } = subEdge;
+              return {
+                ...subEdge,
+                node: {
+                  ...subEdge.node,
+                  ...(node.id === squad.id && { currentMember: user }),
+                },
+              };
+            }),
+          },
+        };
+      }),
+    };
+  };
+
   const { mutateAsync: joinSquad, isPending: isJoiningSquad } = useMutation({
     mutationFn: useJoinSquad({ squad }),
-
     onError: () => {
       displayToast(labels.error.generic);
     },
+    onMutate: () => {
+      queryClient.setQueryData(
+        categoryQueryKey,
+        joinSquadMutation(queryClient.getQueryData(categoryQueryKey)),
```

Comment:
just to avoid undefined values, if `getQueryData` does not return anything we should not `setQueryData`